### PR TITLE
fix(developer): debugger cleanup 🐞

### PR DESCRIPTION
**Pull request by mcdurdin**
Relates to #5013.

Before starting on new functionality in the debugger, I wanted to
cleanup a bunch of bits and pieces, including:

1. Correct handling of text selection and deletion
2. Refactor of character grid at bottom of debug window
3. Deadkey realignment after text changes
4. Correct tracking of cursor position between events
5. Handling a few destruction edge cases

### CODE REVIEW

File: windows/src/developer/TIKE/child/UfrmDebug.pas [hunk 1]
```diff
@@ -570,9 +592,11 @@ procedure TfrmDebug.Run;
 procedure TfrmDebug.ExecuteEvent(n: Integer);
 begin
   memo.ReadOnly := False;
+  memo.Selection := FSavedSelection;
```
Comment (mcdurdin) [inline windows/src/developer/TIKE/child/UfrmDebug.pas:595]: (4) correct tracking of cursor position between events

File: windows/src/global/delphi/comp/KeymanDeveloperDebuggerMemo.pas [hunk 2]
```diff
@@ -28,11 +28,18 @@ interface
 type
   TKeymanDeveloperDebuggerMessageEvent = procedure(Sender: TObject; var Message: TMessage; var Handled: Boolean) of object;
 
+  TMemoSelection = record
```
Comment (mcdurdin) [inline windows/src/global/delphi/comp/KeymanDeveloperDebuggerMemo.pas:31]: All changes in this file provide values for:

(1) Correct handling of text selection and deletion
(3) Deadkey realignment after text changes
(4) correct tracking of cursor position between events

File: windows/src/developer/TIKE/debug/debugdeadkeys.pas [hunk 3]
```diff
@@ -31,12 +31,14 @@   TDeadKeyInfo = class // Kept as a list of active deadkeys
   private
     FDeleted: Boolean;
     FPosition: Integer;
+    FSavedPosition: Integer;
```
Comment (mcdurdin) [inline windows/src/developer/TIKE/debug/debugdeadkeys.pas:34]: Relates to (3) Deadkey realignment after text changes

File: windows/src/developer/TIKE/debug/Keyman.UI.Debug.CharacterGridRenderer.pas [hunk 4]
```diff
@@ -0,0 +1,286 @@
+unit Keyman.UI.Debug.CharacterGridRenderer;
```
Comment (mcdurdin) [inline windows/src/developer/TIKE/debug/Keyman.UI.Debug.CharacterGridRenderer.pas:1]: Yep, all part of (2) Refactor of character grid at bottom of debug window

File: windows/src/developer/TIKE/child/UfrmDebug.pas [hunk 5]
```diff
@@ -99,10 +99,10 @@   TfrmDebug = class(TTikeForm)
     FEvents: TDebugEventList;
     FUIStatus: TDebugUIStatus;
     FUIDisabled: Boolean;
-    LastSelStart: Integer;
 
     { Deadkey member variables }
     FSelectedDeadkey: TDeadKeyInfo;
+    FSavedSelection: TMemoSelection;
```
Comment (mcdurdin) [inline windows/src/developer/TIKE/child/UfrmDebug.pas:105]: Relates to (4) Correct tracking of cursor position between events

File: windows/src/developer/TIKE/child/UfrmDebug.pas [hunk 6]
```diff
@@ -385,6 +386,23 @@ function TfrmDebug.SetKeyEventContext: Boolean;
 begin
   context := km_kbp_state_context(FDebugCore.State);
 
+  if memo.SelLength > 0 then
```
Comment (mcdurdin) [inline windows/src/developer/TIKE/child/UfrmDebug.pas:389]: Relates to (1) Correct handling of text selection and deletion

File: windows/src/developer/TIKE/child/UfrmDebug.pas [hunk 7]
```diff
@@ -630,15 +654,68 @@ procedure TfrmDebug.ExecuteEventRule(n: Integer);
 
 
 procedure TfrmDebug.ExecuteEventAction(n: Integer);
+  type
```
Comment (mcdurdin) [inline windows/src/developer/TIKE/child/UfrmDebug.pas:657]: All this relates to:
(1) Correct handling of text selection and deletion
(3) Deadkey realignment after text changes

File: windows/src/developer/TIKE/child/UfrmDebug.pas [hunk 8]
```diff
@@ -630,15 +654,68 @@ procedure TfrmDebug.ExecuteEventRule(n: Integer);
...
     m, n: Integer;
     dk: TDeadKeyInfo;
+    state: TMemoSelectionState;
   begin
     // Offset is zero-based, but string is 1-based. Beware!
+    state := SaveMemoSelectionState;
     n := memo.SelStart;
     m := n;
 
+    if memo.SelLength > 0 then
```
Comment (mcdurdin) [inline windows/src/developer/TIKE/child/UfrmDebug.pas:708]: Relates to 
(1) Correct handling of text selection and deletion

File: windows/src/developer/TIKE/child/UfrmDebug.pas [hunk 9]
```diff
@@ -743,31 +822,31 @@ procedure TfrmDebug.ExecuteEventAction(n: Integer);
         // Extended Key State
         flag := flag or KEYFLAG_EXTENDED;
 
+      state := SaveMemoSelectionState;
```
Comment (mcdurdin) [inline windows/src/developer/TIKE/child/UfrmDebug.pas:825]: Relates to
(1) Correct handling of text selection and deletion
(3) Deadkey realignment after text changes
(4) Correct tracking of cursor position between events

File: windows/src/developer/TIKE/child/UfrmDebug.pas [hunk 10]
```diff
@@ -1254,116 +1322,32 @@ function TfrmDebug.ShortcutDisabled(Key: Word): Boolean;
...
   if memo.Focused then
   begin
     frmKeymanDeveloper.barStatus.Panels[0].Text := 'Debugger Active';
-    if memo.SelText = ''
-      then ch := Unicode.CopyChar(memo.Text, memo.SelStart-1, 1)
-      else ch := Copy(memo.SelText, 1, 16);
-
-    if ch = ''
-      then frmKeymanDeveloper.barStatus.Panels[2].Text := ''
-      else frmKeymanDeveloper.barStatus.Panels[2].Text := FormatUnicode(ch);
```
Comment (mcdurdin) [inline windows/src/developer/TIKE/child/UfrmDebug.pas:1275]: The character grid provides this data, no need to also have a poorer version of it in the status bar.

**Review by rc-swag — APPROVED** [review]
lgtm 

Did not build it though

**Comment by keyman-server** [comment]
Changes in this pull request will be available for download in [Keyman version 15.0.98-alpha](https://keyman.com/downloads/releases/alpha/15.0.98)

**Pull request merged**